src/chatd.cpp

```diff
@@ -4144,4 +4178,97 @@ bool Message::isValidEmail(const string &buf)
     return regex_match(buf, regularExpresion);
 }
 
+FilteredHistory::FilteredHistory(DbInterface *db)
+    : mDb(db)
+{
+    initVaraiables();
+    mDb->getHistoryNodeIndex(mNewest, mOldest);
+}
+
+void FilteredHistory::addMessage(const Message &msg, bool isNew)
+{
+    Message* message = new Message(msg);
+    if (isNew)
+    {
+        mBuffer.emplace_front(message);
+        mNewest ++;
+        mDb->addAttachmentMessageToHistoryNode(msg, mNewest);
+    }
+    else
+    {
+        if (find(msg.id()) == mBuffer.end())
+        {
+            mBuffer.emplace_back(message);
+            mOldest --;
+            mDb->addAttachmentMessageToHistoryNode(msg, mOldest);
+        }
+    }
+}
+
+void FilteredHistory::deleteMessage(const Message &msg)
+{
+    auto it = find(msg.id());
+    if (it != mBuffer.end())
+    {
+        it->reset(new Message(msg));
```

Comment:
Can we simply update the content and the `ts` of the message rather than create a new one and destroy the older msg?

Reply:
Solved it